src/autoscaler/db/sqldb/binding_sqldb.go

```diff
@@ -0,0 +1,145 @@
+package sqldb
+
+import (
+	"database/sql"
+	"time"
+
+	"code.cloudfoundry.org/lager"
+	_ "github.com/lib/pq"
+
+	"autoscaler/db"
+)
+
+type BindingSQLDB struct {
+	dbConfig db.DatabaseConfig
+	logger   lager.Logger
+	sqldb    *sql.DB
+}
+
+func NewBidingSQLDB(dbConfig db.DatabaseConfig, logger lager.Logger) (*BindingSQLDB, error) {
+	sqldb, err := sql.Open(db.PostgresDriverName, dbConfig.URL)
+	if err != nil {
+		logger.Error("open-binding-db", err, lager.Data{"dbConfig": dbConfig})
+		return nil, err
+	}
+
+	err = sqldb.Ping()
+	if err != nil {
+		sqldb.Close()
+		logger.Error("ping-binding-db", err, lager.Data{"dbConfig": dbConfig})
+		return nil, err
+	}
+
+	sqldb.SetConnMaxLifetime(dbConfig.ConnectionMaxLifetime)
+	sqldb.SetMaxIdleConns(dbConfig.MaxIdleConnections)
+	sqldb.SetMaxOpenConns(dbConfig.MaxOpenConnections)
+
+	return &BindingSQLDB{
+		dbConfig: dbConfig,
+		logger:   logger,
+		sqldb:    sqldb,
+	}, nil
+}
+
+func (bdb *BindingSQLDB) Close() error {
+	err := bdb.sqldb.Close()
+	if err != nil {
+		bdb.logger.Error("Close-binding-db", err, lager.Data{"dbConfig": bdb.dbConfig})
+		return err
+	}
+	return nil
+}
+
+func (bdb *BindingSQLDB) CreateServiceInstance(serviceInstanceId string, orgId string, spaceId string) error {
+	query := "SELECT FROM service_instance WHERE service_instance_id = $1"
+	rows, err := bdb.sqldb.Query(query, serviceInstanceId)
+	if err != nil {
+		bdb.logger.Error("create-service-instance", err, lager.Data{"query": query, "serviceinstanceid": serviceInstanceId, "orgid": orgId, "spaceid": spaceId})
+		return err
+	}
+
+	if rows.Next() {
+		rows.Close()
+		return db.ErrAlreadyExists
+	}
+	rows.Close()
+
+	query = "INSERT INTO service_instance" +
```

Comment:
How about we directly insert into the table without a select? The serviceInstanceId is the primary key, so a duplicated insertion will cause a duplicated error. By this, only one SQL statement is executed.
And the same for deletion, just delete without select. We can check how many rows affected by the delete statement. If the result is 0, so we can know that the record does not exist.

Reply:
For deletion this can work.
For Insertion, primary key duplication error is just text and comparing text is not a good idea as it might change with different versions.